packages/adblocker-content/adblocker.ts

```diff
@@ -81,25 +95,28 @@ export function extractFeaturesFromDOM(roots: Element[]): {
         '[id]:not(html):not(body),[class]:not(html):not(body),[href]:not(html):not(body)',
       ),
     ]) {
+      // Check if this object belongs to processedElements and skip if we already did the job on this object.
+      if (processedElements.has(element)) {
+        continue;
+      }
+
+      // Add this element to the processed list.
+      processedElements.add(element);
+
+      // Any conditions to filter this element out should be placed under this line:
       if (ignoredTags.has(element.nodeName.toLowerCase())) {
         continue;
       }
 
-      // Update ids
-      const id = element.id;
-      if (id) {
-        ids.add(id);
+      // Extract features from this element.
+      if (element.id) {
+        ids.add(element.id);
       }
 
-      // Update classes
-      const classList = element.classList;
-      if (classList) {
-        for (const cls of classList) {
-          classes.add(cls);
-        }
+      for (const classEntry of element.classList) {
+        classes.add(classEntry);
       }
 
-      // Update href
       const href = element.getAttribute('href');
```

Comment:
I think it would be best to keep the old code and comments. We don't change anything around here